File: examples/Python/tornado_ioloop/taskwork.py
#!/usr/bin/env python

"""
synopsis:
    Task worker
    Connects PULL socket to tcp://localhost:5557
    Collects workloads from ventilator via that socket
    Connects PUSH socket to tcp://localhost:5558
    Sends results to sink via that socket
    Author: Lev Givon <lev(at)columbia(dot)edu>
    Modified for async/ioloop: Dave Kuhlman <dkuhlman(at)davekuhlman(dot)org>
usage:
    python taskwork.py
"""

import sys
from functools import partial
import zmq
from zmq.eventloop.future import Context
from zmq.eventloop.ioloop import IOLoop
from tornado import gen


@gen.coroutine
def run_worker(context):
    # Socket to receive messages on
    receiver = context.socket(zmq.PULL)
    receiver.connect("tcp://localhost:5557")
    # Socket to send messages to
    sender = context.socket(zmq.PUSH)
    sender.connect("tcp://localhost:5558")
    # Process tasks forever
    while True:
        s = yield receiver.recv()
        # Simple progress indicator for the viewer
        sys.stdout.write('.')
        sys.stdout.flush()
        # Do the work
        yield gen.sleep(int(s) * 0.001)
        # Send results to sink
        yield sender.send(b'')


@gen.coroutine
def run(loop):
    context = Context()
    yield run_worker(context)


def main():
    args = sys.argv[1:]
    if len(args) != 0:
        sys.exit(__doc__)
    try:
        loop = IOLoop.current()
        loop.run_sync(partial(run, loop, ))
    except KeyboardInterrupt:
        print('\nFinished (interrupted)')


if __name__ == '__main__':
    main()

File: examples/Java/src/main/java/com/imatix/zguide/ex1_9/TaskWork.java
package com.imatix.zguide.ex1_9;

import org.zeromq.ZMQ.*;

import static com.imatix.zguide.commons.ZConstants.URL.TCP_5557;
import static com.imatix.zguide.commons.ZConstants.URL.TCP_5558;
import static org.zeromq.ZMQ.*;

/**
 * Task worker in Java
 * Connects PULL socket to tcp://localhost:5557
 * Collects workloads from ventilator via that socket
 * Connects PUSH socket to tcp://localhost:5558
 * Sends results to sink via that socket
 *
 * @since 1.0
 */

public class TaskWork {

    public static void main(String[] args) throws InterruptedException {

        Context context = context(1);
        //  Socket to receive messages on
        Socket receiver = context.socket(PULL);
        receiver.connect(TCP_5557);
        //  Socket to send messages to
        Socket sender = context.socket(PUSH);
        sender.connect(TCP_5558);
        //  Process tasks forever
        while (!Thread.currentThread().isInterrupted()) {
            String duration = new String(receiver.recv(0)).trim();
            long msec = Long.parseLong(duration);
            //  Simple progress indicator for the viewer
            System.out.flush();
            System.out.println(duration + " . ");
            //  Do the work
            Thread.sleep(msec);
            //  Send results to sink
            sender.send("".getBytes(), 0);
        }
        sender.close();
        receiver.close();
        context.term();
    }

}
